Mathlib/Data/Nat/Find.lean

```diff
@@ -0,0 +1,92 @@
+/-
+Copyright (c) 2016 Microsoft Corporation. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Leonardo de Moura, Jeremy Avigad
+-/
+
+import Batteries.WF
+import Mathlib.Init.Data.Nat.Basic
+
+namespace Nat
+
+/-!
+# Nat.find
+
+If `p` is a (decidable) predicate on `ℕ` and `hp : ∃ (n : ℕ), p n` is a proof that
+there exists some natural number satisfying `p`, then `Nat.find hp` is the
+smallest natural number satisfying `p`. Note that `Nat.find` is protected,
+meaning that you can't just write `find`, even if the `Nat` namespace is open.
+
+The API for `Nat.find` is:
+
+* `Nat.find_spec` is the proof that `Nat.find hp` satisfies `p`.
+* `Nat.find_min` is the proof that if `m < Nat.find hp` then `m` does not satisfy `p`.
+* `Nat.find_min'` is the proof that if `m` does satisfy `p` then `Nat.find hp ≤ m`.
+
+-/
+
+
+section Find
+
+variable {p : ℕ → Prop}
+
+private def lbp (m n : ℕ) : Prop :=
+  m = n + 1 ∧ ∀ k ≤ n, ¬p k
+
+variable [DecidablePred p] (H : ∃ n, p n)
+
+private def wf_lbp : WellFounded (@lbp p) :=
+  ⟨let ⟨n, pn⟩ := H
+    suffices ∀ m k, n ≤ k + m → Acc lbp k from fun a => this _ _ (Nat.le_add_left _ _)
+    fun m =>
+    Nat.recOn m
+      (fun k kn =>
+        ⟨_, fun y r =>
+          match y, r with
+          | _, ⟨rfl, a⟩ => absurd pn (a _ kn)⟩)
+      fun m IH k kn =>
+      ⟨_, fun y r =>
+        match y, r with
+        | _, ⟨rfl, _a⟩ => IH _ (by rw [Nat.add_right_comm]; exact kn)⟩⟩
+
+protected def findX : { n // p n ∧ ∀ m < n, ¬p m } :=
```

Comment:
Pre-existing: this definition needs a doc-string. Would you like to write one now?